pkg/row/copier_eta_history.go

```diff
@@ -0,0 +1,87 @@
+package row
+
+import (
+	"strings"
+	"time"
+)
+
+const (
+	historyIncrements = 1 * time.Hour
+	oldestHistory     = 24 * time.Hour
+)
+
+type copierETA struct {
+	estimate time.Duration
+	asOf     time.Time
+}
+
+type CopierEtaHistory struct {
+	latestEstimate *copierETA
+	etaHistory     []copierETA
+}
+
+func NewCopierEtaHistory() *CopierEtaHistory {
+	return &CopierEtaHistory{
+		etaHistory: make([]copierETA, 0),
+	}
+}
+
+// Store the given duration as an eta with the current time
+func (c *CopierEtaHistory) AddCurrentEstimateAndCompare(estimate time.Duration) string {
+	c.addETA(copierETA{estimate: estimate, asOf: time.Now()})
+	return c.getComparison()
+}
+
+// Store the given eta if history is empty or if the last ETA was >= historyIncrements ago
+func (c *CopierEtaHistory) addETA(eta copierETA) {
+	// If history is empty, store the eta
+	if len(c.etaHistory) == 0 {
+		c.etaHistory = append(c.etaHistory, eta)
+		return
+	}
+
+	// Store whatever eta we get
+	c.latestEstimate = &eta
+
+	// Only store in the history if the last ETA was >= historyIncrements ago
+	lastETA := c.etaHistory[len(c.etaHistory)-1]
+	if time.Since(lastETA.asOf) >= historyIncrements {
+		c.etaHistory = append(c.etaHistory, eta)
+	}
+
+	// Remove old history
+	for len(c.etaHistory) > 0 && time.Since(c.etaHistory[0].asOf) > oldestHistory {
+		c.etaHistory = c.etaHistory[1:]
+	}
+}
+
+// Return a string showing the difference between the latest ETA and the oldest ETA in the format "+30h from 1d ago"
+func (c *CopierEtaHistory) getComparison() string {
+	if c.latestEstimate == nil || len(c.etaHistory) == 0 {
+		return ""
+	}
+
+	oldest := c.etaHistory[0]
+
+	diff := oldest.estimate - c.latestEstimate.estimate
+	ago := c.latestEstimate.asOf.Sub(oldest.asOf)
+
+	relativeDiff := diff - ago
+	diffStr := shortDuration(relativeDiff.Round(time.Minute))
+	if diffStr == "0s" {
+		diffStr = "±0m"
+	}
+
+	return diffStr + " from " + shortDuration(ago.Round(time.Minute)) + " ago"
+}
+
+func shortDuration(d time.Duration) string {
+	s := d.String()
```

Comment:
You can probably use something like `d.Round(time.Hour)`  here instead https://pkg.go.dev/time#Duration.Round

Reply:
That yields strings like `1h0m` instead of `1h`.

Reply:
Ahh I see, nvm.